src/hotspot/share/opto/printinlining.hpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef PRINTINLINING_HPP
+#define PRINTINLINING_HPP
+
+#include "memory/allocation.hpp"
+#include "utilities/ostream.hpp"
+#include "utilities/growableArray.hpp"
+
+class JVMState;
+class ciMethod;
+enum class InliningResult;
+
+class InlinePrinter {
+private:
+  struct IPInlineAttempt : public ArenaObj {
+    IPInlineAttempt(InliningResult result);
+    const InliningResult result;
+    stringStream msg;
+    int bci;
+  };
+
+  class IPInlineSite : public ArenaObj {
+  private:
+    Arena* const _arena;
+    ciMethod* const _method;
+    int const _bci;
+    GrowableArray<IPInlineAttempt*> _attempts;
+    GrowableArray<IPInlineSite*> _children;
+
+  public:
+    /**
+     * Method may be null iff this is the root of the tree.
+     */
```

Comment:
Just a drive-by comment: We usually use "`//` style" method and class comments in the code base. We should probably stick to it for consistency.